### PR TITLE
Implement validators as traits

### PR DESCRIPTION
**Pull request by dnaka91**
This is the re-implementation of validators as traits, rather than a plain closure trait object, as discussed in #54.

There were a few unexpected hurdles during implementation:

- Boxed trait objects are not cloneable easily. Therefore, using `dyn-clone` as helper to make the clone-able. Otherwise, all prompts would lose the ability to be cloned.
- Some prompts have both `with_validator` and `with_validators` functions. The first one could be made generic, calling `Box::new` internally, whereas the second one directly takes a slice of `Box` as a parameter (due to how generics work).
- Closures now seem to always require the parameter type to be defined. This is likely to how the generics are being resolved.

### Todo

Still have to update several docs, just wanted to get some input about this implementation (and its effects on the API) first.

### Future work

I think if this one looks good, we could continue doing the same change for formatters and parsers. What do you think?

### CODE REVIEW

**Comment by mikaelmello** [comment]
>Closures now seem to always require the parameter type to be defined. This is likely to how the generics are being resolved.

IIRC I always ended up needing to explicitly define it on my personal use as well :)

>I think if this one looks good, we could continue doing the same change for formatters and parsers. What do you think?

Completely agree.

---

I'll take a closer (and gradual) look over the next week

File: examples/text_options.rs
Comment (mikaelmello) [inline examples/text_options.rs:14]: I'm also not a huge fan of this syntax, even with simple macros such as `required!()`, the Box thing adds a huge overhead.

Thought of even removing `with_validators` and pointing users to make multiple calls to `with_validator`, but I think that would be bad for people that define their validators somewhere else and just line them up in this call.

Just wanted to express my thoughts, I understand there's not somewhere to go from here =/

File: src/lib.rs
Comment (mikaelmello) [inline src/lib.rs:40]: Having to put `StringValidator` was something I always disliked. This is much better in the majority of cases :)

**Review by mikaelmello — COMMENTED** [review]
Sorry for taking so long, I really appreciate all the effort that went into this PR and really excited to see it through!

Once we merge it, I'll finally release v0.3.0

File: src/validator.rs
Comment (dnaka91) [reply src/validator.rs:266]: That's an excellent idea, and good usage of `Default`/`new`. Also, can get rid of the `Option`, as we technically always have a message.

What do you think about using `Cow<'static, str>` instead of `String`? Would help to avoid allocations in many cases. On the other hand, it might be a  bit of an over-optimizations, as a terminal crate usually doesn't have to worry too much about a few allocations here and there 🤔.

For example:
```rust
use std::borrow::Cow;

struct ValueRequiredValidator {
    message: Cow<'static, str>,
}

impl ValueRequiredValidator {
    pub fn new(message: impl Into<Cow<'static, str>>) -> Self {
        Self {
            message: message.into(),
        }
    }
}

fn main() {
    // This one allocates a string and takes it as before.
    let v = ValueRequiredValidator::new(format!("hello, {}!", "World"));
    assert!(matches!(v.message, Cow::Owned(_)));

    // This one avoids an allocation and takes a reference to the static string.
    // It would be the same case for ValueRequiredValidator::default().
    let v = ValueRequiredValidator::new("hello, World!");
    assert!(matches!(v.message, Cow::Borrowed(_)));
}
```

File: src/validator.rs
Comment (dnaka91) [reply src/validator.rs:212]: That's a good point. Originally, I thought we don't need macros anymore due to having defined structs for everything now. But it definitely is more verbose, so let's put them back in and update.

Also, it means people who already used the macros, won't have to change too much when updating.

File: examples/multiselect.rs
Comment (dnaka91) [reply examples/multiselect.rs:20]: This one, I spend about 2 or 3 days on, trying to get `rustc` to do some auto-dereference for this type, to have `&[ListOption<&str>]` instead of the double reference.

Would have loved for `rustc` to just take it without any extra type annotations even, as the type should be clear from the usage, and even rust-analyzer figures out the right type.

After several days, I gave up and just accepted this unfortunate type declaration. Simply couldn't find any way of making `rustc` infer the type or drop the second reference.

File: examples/text_options.rs
Comment (dnaka91) [reply examples/text_options.rs:14]: Same opinion here.

What do you think about updating all examples to repeatedly call `with_validator` instead? So the API would still exist for people who directly want to pass a pre-created slice, but we show the _preferred_ way in the examples, guiding users to a visually cleaner approach (at the cost of eventual extra allocations due to `Vec` re-sizing, but not a huge concern in my opinion).

**Comment by dnaka91** [comment]
Thank you for this really in-depth review ❤️.

I'm not in a hurry, so no worries about taking time with the review. I'll update some parts according to your comments, and then wait for your reply on the other parts.

File: src/validator.rs
Comment (dnaka91) [reply src/validator.rs:281]: What do you think about the other validators? I believe it may be reasonable to do the same there.

For example, the `MaxLengthValidator` could become:
```rust
#[derive(Clone)]
pub struct MaxLengthValidator {
    limit: usize,
    message: String,
}

impl MaxLengthValidator {
    /// Create a new instance of this validator, requiring at most the given length, otherwise
    /// returning an error with given message.
    ///
    /// If the message is omitted, `The length of the response should be at most $length` is used
    /// as default.
    pub fn new(limit: usize, message: Option<impl Into<String>>) -> Self {
        Self {
            limit,
            message: match message {
                Some(msg) => msg.into(),
                None => format!("The length of the response should be at most {}", limit),
            },
        }
    }
}
```

Comment (dnaka91) [reply src/validator.rs:281]: Actually, better to not do that. If we pass `None` for `message`, then rustc can't infer the type and would require a very _ugly_ function call, where all we want is, simply passing `None` as argument.

File: examples/multiselect.rs
Comment (mikaelmello) [reply examples/multiselect.rs:20]: Yeah that's understandable, maybe at some point rustc will improve in this scenario and we'll get the benefits without any changes.

File: examples/text_options.rs
Comment (mikaelmello) [reply examples/text_options.rs:14]: I like it, please let's do that.

>(at the cost of eventual extra allocations due to Vec re-sizing, but not a huge concern in my opinion)

Say what, let's create the vector with a initial capacity of 5 (or any other value you see reasonable). Something that is small enough for memory usage and large enough for the common use of validators.

File: src/validator.rs
Comment (mikaelmello) [reply src/validator.rs:266]: Leaving the memory optimization aside, I see a lot of value in making this syntax possible:

```
let v = ValueRequiredValidator::new("hello, World!");
```

_Which I just realized also was possible when you used `impl Into<String>`_

What I'm thinking about is that `'static` lifetime and a little confusion that might come with it in certain cases:

```
fn prompt(validator_message: &str) {
    // `validator_message` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
    let v = ValueRequiredValidator::new(validator_message);
    assert!(matches!(v.message, Cow::Borrowed(_)));
}
```

So we have 3 options I guess for the new argument

- `String`:
  - Pro: Straightforward, newcomers specially will never be confused which is one of my goals in making this crate focused on a  good API.
  - Con: Will always require the string to be allocated.
- `Into<String>`: I guess same as above + allowing the use of raw strings. Downside is cloning even strings already owned.
- `Into<Cow<'static, String>>` which looks good regarding the optimizations but has the potential of making the API a little bit more uninviting, specially for newcomers. There's also the fact that down the line this error message will be converted to an owned string either way on `ErrorMessage`, a public struct where I think using `Cow` would bring more complexity to the API.

In the end I'm fine with either Into<String> or String as parameters, and quite not sure if the Cow is worth it, wdyt?

Comment (dnaka91) [reply src/validator.rs:266]: How about we stick with `Into<String>` then?

> Downside is cloning even strings already owned.
Each type automatically implements `Into<X>` for itself, and in case of `String` I believe it's a no-op. So if we pass some sort of owned string, it'll use it without an extra clone.

File: examples/text_options.rs
Comment (dnaka91) [reply examples/text_options.rs:14]: 5 sounds like a good value. After all, it's a `Vec` of `Box`s, so it'll only be 64 bytes allocated for each prompt's validators ([`24 + 5 * 8 = 64`, on 64-bit](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c03436a75072ea226f5883d5384f83f6)).

File: src/validator.rs
Comment (mikaelmello) [reply src/validator.rs:266]: Good point. I had researched before commenting but turns out I mixed Into<String> with ToString, which does make an unnecessary clone...

Let's go with Into<String>

**Review by mikaelmello — COMMENTED** [review]
Looks great to me. I think the only possible remaining change is in regards to the cfg attributes.

Previously, I had guarded the built-in validators behind a feature in order to not pollute everyone with generic-named macros. Right now we could publish the built-in validators without their macro alternatives.

Wdyt of changing the feature from `builtin_validators` to something like `macro_validators` (or perhaps a better name) and only guarding the macros behind it?

**Comment by mikaelmello** [comment]
>I think if this one looks good, we could continue doing the same change for formatters and parsers. What do you think?

I think that would be great. I'll see if I can get started on the formatters over the next week or two. There are some possible additional features in formatters specifically that I might try during the migration. Can I ping you for a review once it's ready?

**Comment by mikaelmello** [comment]
And just to be thorough with the comments and avoid a little bit of back-and-forth, there are also some documentation changes left that are needed right?

From the top of my head we gotta update the changelog and the readme, not sure if there are places in the code still pending.

**Comment by dnaka91** [comment]
> Wdyt of changing the feature from builtin_validators to something like macro_validators (or perhaps a better name) and only guarding the macros behind it?

Sure, no problem. How about `validator-macros` or just `macros`?

> I think that would be great. I'll see if I can get started on the formatters over the next week or two. There are some possible additional features in formatters specifically that I might try during the migration. Can I ping you for a review once it's ready?

Ping me any time 👍. Regarding these changes, I had the idea of using generics here, as it's only a single field instead of a `Vec`. But that turned out to be an awful idea, as it pollutes the type, functions and basically everything with the generics, making it super complex and confusing on the API level. Especially, when introducing more generics for the other fields too.

Just wanted to let you know, that I think trait objects are a great choice here, despite me usually preferring generics (due to the performance  gain).

> And just to be thorough with the comments and avoid a little bit of back-and-forth, there are also some documentation changes left that are needed right?

Exactly, now comes the hardest part, updating and writing documentation 😅. Give me a few days, and I'll go over all the docs, updating where needed.

File: src/prompts/text.rs
Comment (mikaelmello) [inline src/prompts/text.rs:73]: great catches.

**Review by mikaelmello — APPROVED** [review]
Hey I wasn't sure if your last 2 commits were the final updates or if there were more to come.

Just checked and things look great, is it finished?

**Comment by dnaka91** [comment]
Yes, should be ready I think. Hope I didn't forget anything 😅

**Pull request merged**
